service/history/workflow/update/update.go

```diff
@@ -26,85 +26,392 @@ package update
+	opts ...updateOpt,
+) *Update {
+	upd := &Update{
+		id:              id,
+		state:           stateCompleted,
+		instrumentation: &noopInstrumentation,
+		accepted:        future.NewReadyFuture[*failurepb.Failure](nil, nil),
+		outcome:         lazyOutcome(fetchOutcome),
+	}
+	for _, opt := range opts {
+		opt(upd)
 	}
+	return upd
 }
 
+// WaitOutcome observes this Update's completion, returning the Outcome when it
+// is available. This call will block until the Outcome is known or the provided
+// context.Context expires.
 func (u *Update) WaitOutcome(ctx context.Context) (*updatepb.Outcome, error) {
 	return u.outcome.Get(ctx)
 }
 
-func (u *Update) accept() {
-	u.state = stateAccepted
+// WaitAccepted blocks on the acceptance of this update, returning nil if has
+// been accepted but not yet completed or the overall Outcome if the update has
+// been completed (including completed by rejection). This call will block until
+// the acceptance occurs or the provided context.Context expires.
+func (u *Update) WaitAccepted(ctx context.Context) (*updatepb.Outcome, error) {
+	if u.outcome.Ready() {
+		// being complete implies being accepted, return the completed outcome
+		// here because we can.
+		return u.outcome.Get(ctx)
+	}
+	if _, err := u.accepted.Get(ctx); err != nil {
+		return nil, err
+	}
+	return nil, nil
+}
+
+// OnMessage delivers a message to the Update state machine. The proto.Message
+// parameter is expected to be one of *updatepb.Request, *updatepb.Response,
+// *updatepb.Rejection, *updatepb.Acceptance, or a *protocolpb.Message whose
+// Body field contains an instance from the same list. Writes to the EventStore
+// occur synchronously but externally observable effects on this Update (e.g.
+// emmitting an Outcome or an Accepted) are registered with the EventStore to be
+// applied after the durable updates are committed. If the EventStore rolls
+// back its effects, this state machine does the same.
+func (u *Update) OnMessage(
+	ctx context.Context,
+	msg proto.Message,
+	eventStore EventStore,
+) error {
+	if msg == nil {
+		return invalidArgf("Update %q received nil message", u.id)
+	}
+	if protocolMsg, ok := msg.(*protocolpb.Message); ok {
+		var dynbody types.DynamicAny
+		if err := types.UnmarshalAny(protocolMsg.Body, &dynbody); err != nil {
+			return err
+		}
+		msg = dynbody.Message
+	}
+	switch body := msg.(type) {
+	case *updatepb.Request:
+		return u.onRequestMsg(ctx, body, eventStore)
+	case *updatepb.Acceptance:
+		return u.onAcceptanceMsg(ctx, body, eventStore)
+	case *updatepb.Rejection:
+		return u.onRejectionMsg(ctx, body, eventStore)
+	case *updatepb.Response:
+		return u.onResponseMsg(ctx, body, eventStore)
+	default:
+		return invalidArgf("Message type %T not supported", body)
+	}
+}
+
+// PollOutboundMessages loads any oubound messages from this Update state
+// machine into the output slice provided.
+func (u *Update) PollOutboundMessages(out *[]*protocolpb.Message) {
```

Comment:
I don't like word `Poll` here. Poll in Temporal is "wait 1 minute until there is something new".

Reply:
I'll do Read since there's a reasonable analogy to io.Reader